domain/relation/state/relation.go

```diff
@@ -25,3 +33,101 @@ func NewState(factory database.TxnRunnerFactory, clock clock.Clock, logger logge
 		logger:    logger,
 	}
 }
+
+// GetRelationEndpointUUID retrieves the endpoint UUID of a given relation
+// for a specific application.
+// It queries the database using the provided application ID and relation UUID
+// arguments.
+//
+// The following error types can be expected to be returned:
+//   - [relationerrors.ApplicationNotFound] is returned if the application
+//     is not found.
+//   - [relationerrors.RelationEndpointNotFound] is returned is the relation
+//     Endpoint is not found.
+//   - [relationerrors.RelationNotFound] is returned is the relation UUID
+//     is not found.
+func (st *State) GetRelationEndpointUUID(ctx context.Context, args relation.GetRelationEndpointUUIDArgs) (
+	corerelation.EndpointUUID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	type relationEndpointUUID struct {
+		UUID string `db:"uuid"`
+	}
+	type relationEndpointArgs struct {
+		AppID        string `db:"application_uuid"`
+		RelationUUID string `db:"relation_uuid"`
+	}
+	dbArgs := relationEndpointArgs{
+		AppID:        args.ApplicationID.String(),
+		RelationUUID: args.RelationUUID.String(),
+	}
+	stmt, err := st.Prepare(`
+SELECT re.uuid AS &relationEndpointUUID.uuid
+FROM relation_endpoint re
+JOIN application_endpoint ae ON re.endpoint_uuid = ae.uuid
+WHERE ae.application_uuid = $relationEndpointArgs.application_uuid
+AND re.relation_uuid = $relationEndpointArgs.relation_uuid
+`, relationEndpointUUID{}, dbArgs)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	var relationEndpoint relationEndpointUUID
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, dbArgs).Get(&relationEndpoint)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// Check if it is a missing application
+			appFound, err := checkExistsByUUID(ctx, st, tx, "application", args.ApplicationID.String())
+			if err != nil {
+				return errors.Capture(err)
+			}
+			// check if the relation exists
+			relationFound, err := checkExistsByUUID(ctx, st, tx, "relation", args.RelationUUID.String())
+			if err != nil {
+				return errors.Capture(err)
+			}
+			var errs []error
+			if !appFound {
+				errs = append(errs, fmt.Errorf("%w: %s", relationerrors.ApplicationNotFound, args.ApplicationID))
+			}
+			if !relationFound {
+				errs = append(errs, fmt.Errorf("%w: %s", relationerrors.RelationNotFound, args.ApplicationID))
+			}
+			if len(errs) > 0 {
+				return errors.Join(errs...)
+			}
+			return fmt.Errorf("%w: %s", relationerrors.RelationEndpointNotFound, args.RelationUUID)
```

Comment:
todo: Add application.ID to the message. A relation usually has 2 endpoints, one for each application.